### PR TITLE
[WIP] PUT/GET API for object lifecycle management

### PR DESCRIPTION
**Pull request by nguptaopensds**
**What this PR does / why we need it**:
This PR is for PUT/GET API for object lifecycle management

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
NA

**Special notes for your reviewer**:
Review PUT API, GET API is still WIP, so it is not yet updated, this is just a first version of the APIs
PUT 
```
http://127.0.0.1:8089/v1/s3/test/?lifecycle
<LifecycleConfiguration>
  <Rule>
    <ID>id1</ID>
    <Filter>
       <Prefix>documents/</Prefix>
    </Filter>
    <Status>Enabled</Status>
    <Transition>
      <Days>30</Days>
      <StorageClass>GLACIER</StorageClass>
    </Transition>
  </Rule>
  <Rule>
    <ID>id2</ID>
    <Filter>
       <Prefix>logs/</Prefix>
    </Filter>
    <Status>Enabled</Status>
    <Expiration>
      <Days>365</Days>
    </Expiration>
  </Rule>
</LifecycleConfiguration>
```
GET
//TODO
**Release note**:

```release-note
```

### CODE REVIEW

**Comment by nguptaopensds** [comment]
@sfzeng Please review 
@skdwriting  Please review

**Comment by sfzeng** [comment]
1. For api/cmd/main.go:
Why you updated api/cmd/main.go ?

2. For api/pkg/examples/policy.json:
bucketlifecycleput should be bucketlifecycle:put, bucketlifecycleget should be bucketlifecycle:get

3. For api/pk/s3/router.go:
The following does not need to be added, because the RouteBucketGet can route the request to the specific function.
	ws.Route(ws.PUT("/{bucketName}/?lifecycle").To(handler.RouteBucketPut)).Doc("Create lifecycle configuration for the bucket")
	ws.Route(ws.GET("/{bucketName}/?lifecycle").To(handler.RouteBucketGet)).Doc("Get lifecycle configuration from the bucket")

4. For s3/proto/s3.proto:
Your code is conflict with the latest version of developement branch, please pull the latest version and change based on that.

5. For vendor/github.com/golang/protobuf/***:
Why need created so many files? We did not have them, but multi-cloud work well, so please make sure if it is necessary.

File: api/cmd/main.go [hunk 1]
```diff
@@ -45,22 +44,23 @@ func main() {
 	ws.Doc("OpenSDS Multi-Cloud API")
 	ws.Consumes(restful.MIME_JSON)
 	ws.Produces(restful.MIME_JSON)
-	ws.Filter(auth.FilterFactory())
+	//ws.Filter(auth.FilterFactory())
```
Comment (wisererik) [inline api/cmd/main.go]: should revert main.go file

File: api/pkg/s3/bucketlifecycleget.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2018 Huawei Technologies Co., Ltd. All Rights Reserved.
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleget.go]: copyright is not correct

File: api/pkg/s3/bucketlifecycleget.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+import (
+	"github.com/emicklei/go-restful"
+	"github.com/micro/go-log"
+	"github.com/opensds/multi-cloud/api/pkg/policy"
+	"net/http"
+	"encoding/json"
+	"fmt"
+	"github.com/opensds/multi-cloud/api/pkg/common"
+	//	"github.com/micro/go-micro/errors"
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleget.go]: remove unused library

File: api/pkg/s3/bucketlifecycleget.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		log.Logf("Get filter failed: %v", err)
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	} else {
+		log.Logf("Get filter for BucketGet, filterOpts=%+v, filter=%+v\n",
+			filterOpts, filter)
+	}
+
+	if filter[common.KObjKey] != "" {
+		//filter[common.KObjKey] should be like: like:parttern
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleget.go]: need space after //

File: api/pkg/s3/bucketlifecycleput.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2018 Huawei Technologies Co., Ltd. All Rights Reserved.
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleput.go]: copyright is not correct

File: api/pkg/s3/bucketlifecycleput.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+import (
+	"encoding/xml"
+	"net/http"
+
+	"github.com/emicklei/go-restful"
+	"github.com/micro/go-log"
+
+	"github.com/opensds/multi-cloud/api/pkg/policy"
+	//. "github.com/opensds/multi-cloud/s3/pkg/exception"
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleput.go]: remove unused library

File: api/pkg/s3/bucketput.go [hunk 7]
```diff
@@ -35,7 +35,7 @@ func (s *APIService) BucketPut(request *restful.Request, response *restful.Respo
 		return
 	}
 	bucketName := request.PathParameter("bucketName")
-
+        //This is neelam comment
```
Comment (wisererik) [inline api/pkg/s3/bucketput.go]: remove or update this comment

File: api/pkg/s3/routbucketget.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2018 Huawei Technologies Co., Ltd. All Rights Reserved.
```
Comment (wisererik) [inline api/pkg/s3/routbucketget.go]: copyright is not correct

File: s3/proto/s3.proto [hunk 9]
```diff
@@ -135,7 +154,7 @@ message Bucket {
 	WebsiteConfiguration websiteConfiguration = 11;
 	CORSConfiguration cors = 12;
 	ReplicationConfiguration replicationConfiguration = 13;
-	repeated LifecycleRole lifecycleConfiguration = 14;
+	repeated LifecycleRule LifecycleConfiguration = 14;
```
Comment (wisererik) [inline s3/proto/s3.proto]: first letter of LifecycleConfiguration is lowercase

File: s3/pkg/model/xmlstruct.go [hunk 10]
```diff
@@ -79,3 +79,36 @@ type ListPartsOutput struct {
 	Owner       Owner  `xml:"Owner"`
 	Parts       []Part `xml:"Part"`
 }
+
+type LifecycleConfiguration struct {
+	Rule        []Rule   	`xml:"Rule"`
+}
+
+type Rule struct {
```
Comment (wisererik) [inline s3/pkg/model/xmlstruct.go]: should format this xmlstruct.go file

**Comment by yuejuSDS** [comment]
while creating lifeCycle, It's necessary to verify that the <id> can't be duplicated, please add <id> check.

**Comment by wisererik** [comment]
please fix CI error

File: s3/proto/s3.proto (hunk 9 above)
Comment (leonwanghui) [reply s3/proto/s3.proto]: s/LifecycleConfiguration/lifecycleConfiguration

**Comment by leonwanghui** [comment]
Also please reslove the conflict errors.

File: api/pkg/s3/bucketlifecycleput.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+func (s *APIService) BucketLifecyclePut(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "bucket:put") {
+		return
+	}
+	bucketName := request.PathParameter("bucketName")
+	log.Logf("Received request for create bucket: %s", bucketName)
+	ctx := context.Background()
+	bucket, _ := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
+	body := ReadBody(request)
+	log.Logf("Body request is %v\n", body)
```
Comment (kumarashit) [inline api/pkg/s3/bucketlifecycleput.go]: you can merge the two logs into one

File: s3/proto/s3.proto (hunk 9 above)
Comment (sfzeng) [reply s3/proto/s3.proto]: Why replace Action with ExpirationAction and TransitionAction ? Could you give some explanation here?

File: api/pkg/s3/bucketlifecycleput.go (hunk 11 above)
Comment (sfzeng) [reply api/pkg/s3/bucketlifecycleput.go]: >> res, err := s.s3Client.CreateBucket(ctx, bucket) 
When create lifecycle, only need to update the lifecycle configuration of a bucket, so it's no good to use CreateBucket.

**Comment by nguptaopensds** [comment]
link to new PR : https://github.com/opensds/multi-cloud/pull/127

**Pull request closed without merging**